File: users/keshengqiang/684.冗余连接.cpp
class Solution {
public:
	vector<int> parent;
	void Init(int len) {
		parent.resize(len + 1);
		for (int i = 1; i < len; i++) {
			parent[i] = i;
		}
	}
    int Find(int i){
        if(parent[i] == i){
            return i;
        }
        parent[i] = Find(parent[i]);
        return parent[i];
    }

    void Union(int a,int b){
        parent[Find(a)] = Find(b);
    }
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        Init(edges.size());

        for(int i = 0; i < edges.size(); i++) {
            if(Find(edges[i][0]) == Find(edges[i][1])){
                return edges[i];
            } else {
                Union(edges[i][0], edges[i][1]);
            }
        }
        return {};
    }
};

File: users/chenjiasheng/684. 冗余连接.py
from typing import List


class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        # init
        parents = list(range(len(edges)))
        sizes = [1] * len(edges)

        def find(x):
            if parents[x] != x:
                parents[x] = find(parents[x])
            return parents[x]

        def union(x, y):
            x, y = find(x), find(y)
            if x == y:
                return False
            else:
                if sizes[x] < sizes[y]:
                    x, y = y, x
                parents[y] = x
                sizes[x] += sizes[y]
                sizes[y] = 0
                return True

        for x, y in edges:
            if not union(x-1, y-1):
                return [x, y]


edges = [[1, 2], [2, 3], [3, 4], [1, 4], [1, 5]]
solution = Solution()
result = solution.findRedundantConnection(edges)
print(result)

